google-cloud-storage/src/test/java/com/google/cloud/storage/ITJsonResumableSessionPutTaskTest.java

```diff
@@ -0,0 +1,807 @@
+        TmpFile tmpFile =
+            DataGenerator.base64Characters().tempFile(temp.newFolder().toPath(), _256KiBL)) {
+      URI endpoint = fakeHttpServer.getEndpoint();
+      String uploadUrl = String.format("%s/upload/%s", endpoint.toString(), UUID.randomUUID());
+
+      AtomicLong confirmedBytes = new AtomicLong(-1L);
+
+      JsonResumableSessionPutTask task =
+          new JsonResumableSessionPutTask(
+              httpClientContext,
+              uploadUrl,
+              RewindableHttpContent.of(tmpFile.getPath()),
+              HttpContentRange.of(ByteRangeSpec.explicit(0L, _256KiBL)));
+
+      StorageException se = assertThrows(StorageException.class, () -> task.call());
+      assertThat(se.getCode()).isEqualTo(400);
+      assertThat(se.getReason()).isEqualTo("invalid");
+      assertThat(confirmedBytes.get()).isEqualTo(-1L);
+    }
+  }
+
+  /**
+   *
+   *
+   * <h2>S.2</h2>
+   *
+   * Attempting to finalize a session with fewer bytes than GCS acknowledges.
+   *
+   * <table>
+   *   <tr>
+   *     <td>server state</td>
+   *     <td><pre>
+   * persisted_size = 524288
+   *     </pre></td>
+   *   </tr>
+   *   <tr>
+   *     <td>client state</td>
+   *     <td><pre>
+   * write_offset = 262144, finish = true
+   *     </pre></td>
+   *   </tr>
+   *   <tr>
+   *     <td>request</td>
+   *     <td><pre>
+   * PUT $UPLOAD_ID
+   * Content-Range: bytes *&#47;262144
+   *     </pre></td>
+   *   </tr>
+   *   <tr>
+   *     <td>response</td>
+   *     <td><pre>
+   * 308 Resume Incomplete
+   * Range: bytes=0-524287
+   *     </pre></td>
+   *   </tr>
+   * </table>
+   */
+  @Test
+  public void scenario2() throws Exception {
+
+    HttpRequestHandler handler =
+        req -> {
+          FullHttpResponse resp =
+              new DefaultFullHttpResponse(req.protocolVersion(), RESUME_INCOMPLETE);
+          ByteRangeSpec range = ByteRangeSpec.explicit(0L, _512KiBL);
+          resp.headers().set(HttpHeaderNames.RANGE, range.getHttpRangeHeader());
+          return resp;
+        };
+
+    try (FakeHttpServer fakeHttpServer = FakeHttpServer.of(handler)) {
+      URI endpoint = fakeHttpServer.getEndpoint();
+      String uploadUrl = String.format("%s/upload/%s", endpoint.toString(), UUID.randomUUID());
+
+      AtomicLong confirmedBytes = new AtomicLong(-1L);
+
+      JsonResumableSessionPutTask task =
+          new JsonResumableSessionPutTask(
+              httpClientContext,
+              uploadUrl,
+              RewindableHttpContent.empty(),
+              HttpContentRange.of(_256KiBL));
+
+      StorageException se = assertThrows(StorageException.class, () -> task.call());
+      assertThat(se.getCode()).isEqualTo(400);
+      assertThat(se.getReason()).isEqualTo("invalid");
+      assertThat(confirmedBytes.get()).isEqualTo(-1L);
+    }
+  }
+
+  /**
+   *
+   *
+   * <h2>S.3</h2>
+   *
+   * Attempting to finalize a session with more bytes than GCS acknowledges.
```

Comment:
GCS expects instead of acknowledges?

Reply:
Not expects, acknowledges. The response headers is `Range: bytes=0-262143`, which is the number of bytes that have been acknowledged.